### DISCUSSION TITLE
how can I define the type result of "make_storage"  function as a member variable of c++ class.

### DISCUSSION THREAD

**Discussion by CodingPanda87** [fnc12/sqlite_orm#786] [Q&A] [answered]
auto storage = make_storage("db.sqlite",
                            make_table("users",
                                       make_column("id", &User::id, autoincrement(), primary_key()),
                                       make_column("first_name", &User::firstName),
                                       make_column("last_name", &User::lastName),
                                       make_column("birth_date", &User::birthDate),
                                       make_column("image_url", &User::imageUrl),
                                       make_column("type_id", &User::typeId)),
                            make_table("user_types",
                                       make_column("id", &UserType::id, autoincrement(), primary_key()),
                                       make_column("name", &UserType::name, default_value("name_placeholder"))));

for example:

class myBusiness
{
public:
        bool initial(std::string filePath){
               m_storage = make_storage(filePath,
                            make_table("users",
                                       make_column("id", &User::id, autoincrement(), primary_key()),
                                       make_column("first_name", &User::firstName),
                                       make_column("last_name", &User::lastName),
                                       make_column("birth_date", &User::birthDate),
                                       make_column("image_url", &User::imageUrl),
                                       make_column("type_id", &User::typeId)),
                            make_table("user_types",
                                       make_column("id", &UserType::id, autoincrement(), primary_key()),
                                       make_column("name", &UserType::name, default_value("name_placeholder"))));
               return m_storage.is_open();
         }
private:
        xxxxx_type m_storage;
}

what is "xxxxx_type"?

thank you very much  !!!

**Comment by fnc12** [comment, marked as answer]
https://github.com/fnc12/sqlite_orm/wiki/FAQ

**Reply by CodingPanda87** [reply]
thank you!

**Comment by DominicD** [comment]
@fnc12 
I also have a problem with this. I tried this:

        #include "sqlite_orm/sqlite_orm.h"
        
        struct User
        {
	        int id;
        };
        
        inline auto initStorage(const std::string& path) {
	        using namespace sqlite_orm;
	        return
		        make_storage(path,
			        make_table("users",
				        make_column("id", &User::id, autoincrement(), primary_key())
			        )
		        );
        }
        using Storage = decltype(initStorage(""));
        
        class Example
        {
        public:
	        bool initial(std::string filePath) {
		        m_storage = initStorage(filePath);
	        }
        private:
        
	        Storage m_storage;
        };

This gives me following error for "m_storage = initStorage(filePath);"

error C2280: 'sqlite_orm::internal::storage_t<sqlite_orm::internal::table_t<User,sqlite_orm::internal::column_t<User,int,const int &(__cdecl User::* )(void) const,void (__cdecl User::* )(int),sqlite_orm::constraints::autoincrement_t,sqlite_orm::constraints::primary_key_t<>>>> &sqlite_orm::internal::storage_t<sqlite_orm::internal::table_t<User,sqlite_orm::internal::column_t<User,int,const int &(__cdecl User::* )(void) const,void (__cdecl User::* )(int),sqlite_orm::constraints::autoincrement_t,sqlite_orm::constraints::primary_key_t<>>>>::operator =(const sqlite_orm::internal::storage_t<sqlite_orm::internal::table_t<User,sqlite_orm::internal::column_t<User,int,const int &(__cdecl User::* )(void) const,void (__cdecl User::* )(int),sqlite_orm::constraints::autoincrement_t,sqlite_orm::constraints::primary_key_t<>>>> &)': attempting to reference a deleted function


But even if it worked. I actually want something else.
 I don´t want to expose sqlite_orm outside of this class so I do not want to have the include in the header. Same goes for everything that is used in initStorage (in this case the User struct). However seems like a inline function can not be defined in the cpp if it is declared in the header.

Only thing I came up with is this:

        // example.h :
        #include <any>
        #include <string>
        
        class Example
        {
        public:
	        void initial(std::string filePath);
        private:
        
	        std::any m_storage;
        };
        
        //example.cpp :
        //#include "example.h"
        #include "sqlite_orm/sqlite_orm.h"
        
        struct User
        {
	        int id;
        };
        
        inline auto initStorage(const std::string& path) {
	        using namespace sqlite_orm;
	        return
		        make_storage(path,
			        make_table("users",
				        make_column("id", &User::id, autoincrement(), primary_key())
			        )
		        );
        }
        using Storage = decltype(initStorage(""));
        
        void Example::initial(std::string filePath) {
	        m_storage = initStorage(filePath);
        
	        // to use it I have to make ugly cast:
	        auto user = std::any_cast<Storage&>(m_storage).get<User>();
        }


Would be great if you have a better solution. Maybe I am missing something obvious?

**Reply by fnc12** [reply]
you cannot assign a storage to a storage. You need to init a storage (to call a constructor). You you have a later init then store your storage as a nullable value:
instead of `Storage m_storage;` store `std::unique_ptr<Storage> m_storage;`

**Reply by DominicD** [reply]
Ah yes this solves the first example.
But what about the second one? Which I actually want to use.
Any advice how it could be improved?

**Reply by fnc12** [reply]
@DominicD which example are you talking about? How to eliminate includes of your data files inside header? If so then it is unable cause storage uses all these types inside its type declaration just like `std::tuple` uses all types included as template argument. You can try to use forward declaration but it requires using of `sqlite_orm` private API. The example also exists inside FAQ https://github.com/fnc12/sqlite_orm/wiki/FAQ#slow--size-effective

**Reply by DominicD** [reply]
Ok thank you for clarification
